data-connection-jdbc/src/main/java/io/micronaut/data/connection/jdbc/operations/OracleConnectionClientInfoCustomizer.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright 2017-2024 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.data.connection.jdbc.operations;
+
+import io.micronaut.context.annotation.Requires;
+import io.micronaut.core.util.StringUtils;
+import io.micronaut.data.connection.ConnectionDefinition;
+import io.micronaut.data.connection.support.ConnectionClientInfoDetails;
+import jakarta.inject.Singleton;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.sql.Connection;
+import java.sql.SQLClientInfoException;
+import java.sql.SQLException;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * A customizer for Oracle database connections that sets and clears client information.
+ *
+ * This customizer checks if the connection is an Oracle database connection and then sets the client information
+ * (client ID, module, and action) after opening the connection. It also clears these properties before closing the connection.
+ *
+ * @since 4.10
+ */
+@Singleton
+@Requires(property = ConnectionCustomizer.PREFIX + ".oracleclientinfo.enabled", value = "true", defaultValue = "false")
```

Comment:
this I think should use `@EachBean` and be configurable per data source. You can create another `@EachProperty("datasources")` and then the setting can be:

```
datasources.whatever.enable-oracleclientinfo=true
```

We can also probably default to `true`

Reply:
I had in mind dynamic creation of datasources we are seeing recently people need and this is why I wanted to avoid linking to the datasource config, but it does make sense to declare this per datasource.

Reply:
Added `@EachBean(DataSource)` with `@Required(condition)` with custom condition that checks oracle dialect and flag enabled (not enabled by default). `@EachProperty(value = "datasource"...)` was not creating listener instance(s). Now, if users generate `DataSource` dynamically I guess these might get created, but because of condition checking configuration properties then these will not be created. Maybe we should change this class to be public so users can create them via factory if needed.